src/pybasic/pybasic.py

```diff
@@ -0,0 +1,115 @@
+"""Main BaSiC class."""
+
+import os
+from typing import Iterable, List, Union
+
+# import jax
+import jax.numpy as jnp
+import numpy as np
+
+from .profile import Profile
+from .settings import Settings
+
+# from pybasic.tools.dct2d_tools import dct2d, idct2d
+# from pybasic.tools.inexact_alm import inexact_alm_rspca_l1
+
+PathLike = Union[str, bytes, os.PathLike]
+
+
+mm = jnp.matmul
+
+
+# NOTE how should multiple channels be handled?
+# NOTE if settings need to be updated, should user reinitialize class?
+class BaSiC:
+    """A class for fitting and applying BaSiC illumination correction profiles."""
+
+    def __init__(
+        self,
+        settings: Settings,
+        device: str = "cpu",
+    ) -> None:
+        """Inits the BaSiC class with the provided settings.
+
+        Args:
+            settings: :meth:`pybasic.Settings` object
+            device: device to use, options are `"cpu"`, `"gpu"`, `"tpu"`
+        """
+        self._settings = settings
+        self._device = device
+        ...
+
+    def fit(self, images: Iterable[np.ndarray]) -> List[Profile]:
+        """Generate illumination correction profiles.
+
+        Args:
+            images: input images to predict illumination model
+
+        Returns:
+            correction profile model :meth:`pybasic.Model` objects
+
+        Example:
+            >>> from pybasic import BaSiC, Settings
+            >>> from pybasic.tools import load_images
+            >>> images = load_images('./images', lazy=True)
+            >>> settings = Settings()
+            >>> basic = BaSiC(settings)
+            >>> profiles = basic.fit(images)
+
+        Tip:
+            * Use a generator to provide images, reducing memory usage
+        """
+        settings = self._initialize_settings(images)
+        return self._run(images, settings)
```

Comment:
Following the scheme of scikit learn, I think this should not return anything. If we consider PyBaSiC and ML model, the flatfield, darkfield, and photobleaching correction factors are the "weights" of the model, and we should store them internally.

We can create functions to "dump the model" weights to disk in the form of some of standard image format. We should also store the "model score" internally and implement a `score()` method so we can take a look at the final score if someone needs to do debugging.